doc/developers/cython.rst

```diff
@@ -0,0 +1,158 @@
+
+         git clone https://github.com/python/cpython.git
+         conda create -n debug-scikit-dev
+         conda activate debug-scikit-dev
+         cd cpython
+         mkdir debug
+         cd debug
+         ../configure --prefix=$CONDA_PREFIX --with-pydebug
+         make EXTRA_CFLAGS='-DPy_DEBUG' -j<num_cores>
+         make install
+
+  * You might find this alias handy:
+
+    .. code-block::
+
+         alias cythonX="cython -X language_level=3 -X boundscheck=False -X wraparound=False -X initializedcheck=False -X nonecheck=False -X cdivision=True"
+
+  * which you can use with:
+
+    .. code-block::
+
+         cythonX --annotate source.pyx
+
+Tips for performances
+^^^^^^^^^^^^^^^^^^^^^
+
+* Understand the GIL in context for CPython (which problems it solves, what are its limitations) and get a good
+  understanding of when Cython will be mapped to C code free of interactions with CPython, when it will not, and when
+  it cannot (e.g. presence of interactions with Python objects, which include functions). In this regard,
+  `PEP073 <https://peps.python.org/pep-0703/>`_ provides a good overview and context and pathways for removal.
+
+* Make sure you have deactivated `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_.
+
+* Always prefer memoryviews instead over ``cnp.ndarray`` when possible: memoryview are lightweight.
+
+* Avoid memoryview slicing: memoryview slicing might be costly or misleading in some cases and we better not use it IMO,
+  even if handling fewer dimensions in some context would be preferable.
+
+* Decorate final classes or methods with ``@final`` (this allows removing virtual tables when needed)
+
+* Inline methods and function when it makes sense
+
+* Make sure your Cython compilation units `use NumPy recent C API <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/setup.py#L64-L70>`_.
+
+* In doubt, read the generated C or C++ code if you can: "The fewer C instructions and indirections for a line of
+  Cython code, the better" is a good rule of thumb.
+
+Cython Internals
+----------------
+
+Concrete method dispatch implementation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* In most cases, concrete method dispatch is performed **at runtime using virtual tables**
+  (i.e. static C-array of pointers to a given types of function, used at runtime to known which
+  function to call from an instance). This is costly because there is a double indirection.
+
+* In some cases, concrete method dispatch is performed **statically** (a function is called
+  directly by a pointer dereference). This is less costly, especially if there are rooms
+  for this function to be inline.
+
+Known Limitations
+^^^^^^^^^^^^^^^^^
+
+* No template for types exist in Cython, alternatives are:
+
+  * using fused-types
+  * using Tempita, a small preprocessing language to expand Cython source
```

Comment:
Is it worth mentioning C-pointer casting as well? If it's a struct in the function signature. Or too inelegant?

https://www.codementor.io/@arpitbhayani/powering-inheritance-in-c-using-structure-composition-176sygr724

Reply:
I am not sure I understand.

How is it a limitation? Moreover, how would you phrase it?

Reply:
Using fused-types is not possible if say the struct is a extension type attribute:

```Cython
cdef struct AStruct:
     double member

cdef class A:
    AStruct B

    cdef do_something(self):
          B.member = 0.0
```

So instead, if one wants to subclass A and add members to `B`, then one would need to either use tempita, or perform C-casting.

```Cython
cdef struct newAStruct:
     Astruct origstruct
     double newmember

cdef class newA(A):
    AStruct B

    cdef do_something(self):
          B = <newAStruct> self.B
          B.member = 0.0
          B.newmember = 0.0
```

Reply:
Have you found places where such a pattern has been used or would be useful?

Reply:
I used it in the Oblique splitter PR #22754. Since it's not merged, not sure if it's warranted to put the idea onto the Cython doc. I proposed the thought to provide some extra Cython ideas. Feel free to veto it :p

Reply:
I guess we better add this then if such changes get into scikit-learn's codebase. Could you create a pull request in this case?